From 0fe922f23b8692b34ec9ab1925d710ee2948cf5e Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Wed, 29 Nov 2023 16:34:41 +0200
Subject: [PATCH] [FSW-334] fs1: implement image start

image_start() function is implemented for FS1 device.
since ppboot is expected to jump only to preboot, and
preboot is planned to run on spsram, the address translation
function supports only spsram for the moment. other memory regions
will be added later if needed.

Change-Id: I57c67933bcb4bef61577d56078e0685b415e4908
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 7a6a154..b6fdfb6 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -27,6 +27,7 @@
 #include <device.h>
 #include <fs1_reg_map.h>
 #include <fs1_fw_if.h>
+#include <uart.h>
 #include "common.h"
 #include "hl.h"
 
@@ -239,7 +240,31 @@
 	/* TODO: fill Pcie link offload to ARC0 */
 }
 
+#define SPSRAM_ADDR_MASK	0xFFFFF
+
+static u32 fs1_translate_arc_addr(u64 addr)
+{
+	if (addr >= HL_SP_SRAM_BASE && addr < (HL_SP_SRAM_BASE + (u64)HL_SP_SRAM_SIZE)) {
+		addr = (addr & SPSRAM_ADDR_MASK) + SP_SRAM_BASE;
+		goto ret_addr;
+	}
+
+	preboot_puts_hex64("warn: no address translation: ", addr, true);
+
+ret_addr:
+	return (u32)addr;
+}
+
 void image_start(u64 loadaddr)
 {
-	/* TODO: fill jump to preboot */
+	u32 arc_addr = fs1_translate_arc_addr(loadaddr);
+	void (*image_entry)(void);
+
+	/*
+	 * Jump to __reset of zephyr. After image_entry, control is not
+	 * expected to return here and continue.
+	 */
+	image_entry = (void *)(*(u32 *)arc_addr);
+
+	image_entry();
 }
